File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        
        def create_graph():
            g = defaultdict(list)
            for a, b in edges:
                g[a].append(b)
                g[b].append(a)
            
            return g
        
        
        def bfs(graph):
            q = deque([source])
            visited = [True] * n
            while(q):
                node = q.popleft()
                if(node == destination):
                    return True
                
                for v in graph[node]:
                    if(visited[v]):
                        q.append(v)
                        visited[v] = False
            
            return False
                    
        graph = create_graph()
        return bfs(graph)
                
            

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.go
type DisjointSet struct {
    Parent map[int]int
}

func (ds *DisjointSet) Find(x int) int{
    if(ds.Parent[x] != x) {
        ds.Parent[x] = ds.Find(ds.Parent[x])
    }
    return ds.Parent[x]
}

func (ds *DisjointSet) Union(x, y int) {
    px, py := ds.Find(ds.Parent[x]), ds.Find(ds.Parent[y])
    if px != py {
        ds.Parent[py] = px
    }
}

func NewDisjointSet(elements []int) *DisjointSet {
    ret := &DisjointSet{Parent: make(map[int]int)}
    for _, e := range elements {
        ret.Parent[e] = e
    }
    return ret   
}

func validPath(n int, edges [][]int, source int, destination int) bool {
    elements := make([]int, n)
    for i := 0; i < n; i++ {
        elements[i] = i
    }
    ds := NewDisjointSet(elements)
    for _, edge := range edges {
        ds.Union(edge[0], edge[1])
    }
    return ds.Find(source) == ds.Find(destination)
}